File: backend/app/schemas/responses.py
from typing import Literal

from pydantic import BaseModel, ConfigDict

from app.schemas.enums import ActionType, CardType, GamePhase, PlayerStatus, RoomStatus


class SocketResponseModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class RoomCreateResponse(SocketResponseModel):
    roomId: str
    roomCode: str
    playerId: str
    playerSessionId: str


class RoomJoinResponse(SocketResponseModel):
    roomId: str
    roomCode: str
    playerId: str
    playerSessionId: str


class RoomPlayer(SocketResponseModel):
    playerId: str
    nickname: str
    isReady: bool
    isHost: bool
    status: PlayerStatus


class RoomUpdatedEvent(SocketResponseModel):
    roomId: str
    roomCode: str
    status: RoomStatus
    players: list[RoomPlayer]


class GameStartedEvent(SocketResponseModel):
    roomId: str
    currentPlayerId: str
    turnNumber: int


class TurnStartedEvent(SocketResponseModel):
    currentPlayerId: str
    pendingDraws: int
    turnNumber: int


class PublicGamePlayer(SocketResponseModel):
    playerId: str
    nickname: str
    handCount: int
    status: PlayerStatus


class RecentAction(SocketResponseModel):
    actorPlayerId: str
    actionType: ActionType
    targetPlayerId: str | None = None
    summary: str


class PublicGameStateEvent(SocketResponseModel):
    roomId: str
    phase: GamePhase
    currentPlayerId: str
    pendingDraws: int
    turnNumber: int
    players: list[PublicGamePlayer]
    discardTopCardType: CardType | None
    discardCount: int
    winnerPlayerId: str | None
    recentAction: RecentAction | None


class PrivateCardView(SocketResponseModel):
    cardId: str
    cardType: CardType


class PlayerPrivateStateEvent(SocketResponseModel):
    playerId: str
    hand: list[PrivateCardView]
    visibleFutureCards: list[CardType] | None


class PlayerEliminatedEvent(SocketResponseModel):
    playerId: str
    eliminatedBy: Literal["exploding_kitten"]


class GameEndedEvent(SocketResponseModel):
    winnerPlayerId: str


class ErrorEvent(SocketResponseModel):
    code: str
    message: str
    requestId: str | None = None

File: shared/contracts/responses.ts
import type { ActionType, CardType, GamePhase, PlayerStatus, RoomStatus } from "./enums";

export type RoomCreateResponse = {
  roomId: string;
  roomCode: string;
  playerId: string;
  playerSessionId: string;
};

export type RoomJoinResponse = {
  roomId: string;
  roomCode: string;
  playerId: string;
  playerSessionId: string;
};

export type RoomPlayer = {
  playerId: string;
  nickname: string;
  isReady: boolean;
  isHost: boolean;
  status: PlayerStatus;
};

export type RoomUpdatedEvent = {
  roomId: string;
  roomCode: string;
  status: RoomStatus;
  players: RoomPlayer[];
};

export type GameStartedEvent = {
  roomId: string;
  currentPlayerId: string;
  turnNumber: number;
};

export type TurnStartedEvent = {
  currentPlayerId: string;
  pendingDraws: number;
  turnNumber: number;
};

export type PublicGamePlayer = {
  playerId: string;
  nickname: string;
  handCount: number;
  status: PlayerStatus;
};

export type RecentAction = {
  actorPlayerId: string;
  actionType: ActionType;
  targetPlayerId?: string;
  summary: string;
};

export type PublicGameStateEvent = {
  roomId: string;
  phase: GamePhase;
  currentPlayerId: string;
  pendingDraws: number;
  turnNumber: number;
  players: PublicGamePlayer[];
  discardTopCardType: CardType | null;
  discardCount: number;
  winnerPlayerId: string | null;
  recentAction: RecentAction | null;
};

export type PrivateCardView = {
  cardId: string;
  cardType: CardType;
};

export type PlayerPrivateStateEvent = {
  playerId: string;
  hand: PrivateCardView[];
  visibleFutureCards: CardType[] | null;
};

export type PlayerEliminatedEvent = {
  playerId: string;
  eliminatedBy: "exploding_kitten";
};

export type GameEndedEvent = {
  winnerPlayerId: string;
};

export type ErrorEvent = {
  code: string;
  message: string;
  requestId?: string;
};
